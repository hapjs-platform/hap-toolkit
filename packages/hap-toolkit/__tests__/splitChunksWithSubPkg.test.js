/*
 * Copyright (c) 2021-present, the hapjs-platform Project Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

const fs = require('fs')
const cp = require('child_process')
const glob = require('glob')
const path = require('path')
const del = require('del')
const JSZip = require('jszip')
const { copyApp } = require('hap-dev-utils')
const { compile } = require('../lib')

describe('split chunks on a project', () => {
  const platform = 'native'
  let projectRoot
  let buildDir
  beforeAll(async () => {
    const testAppDir = path.resolve(__dirname, '../fixtures/splitChunks-subPkg-app')
    projectRoot = await copyApp(testAppDir)
    cp.execSync('npm install --registry=https://registry.npmmirror.com/', { cwd: projectRoot })
    buildDir = path.resolve(projectRoot, 'build')
  })

  // 公共js结合分包兼容卡片
  it('split chunks with subpackages', async () => {
    const mode = 'dev'
    // 独立拥有完整的共用js内容
    // 非独立包只拥有该分包引用的共用js内容
    const { stats } = await compile(platform, mode, false, {
      cwd: projectRoot,
      splitChunksMode: 'SMART'
    })
    expect(stats.hasErrors()).toBe(false)
    const result = glob.sync('**/*.json', {
      cwd: buildDir
    })
    expect(result).toMatchSnapshot()
    // 测试分包内容
    const pkgs = [
      { name: 'com.application.demo.rpk', jsonPath: 'app-chunks.json', pageChunks: '' },
      { name: 'com.application.demo.rpk', jsonPath: 'page-chunks.json', pageChunks: '' },
      { name: 'com.application.demo.base.srpk', jsonPath: 'page-chunks.json', pageChunks: '' },
      { name: 'com.application.demo.base.srpk', jsonPath: 'app-chunks.json', pageChunks: '' },
      {
        name: 'com.application.demo.sub2.srpk',
        jsonPath: 'Sub2/page-chunks.json',
        pageChunks: ''
      },
      { name: 'com.application.demo.sub3.srpk', jsonPath: 'Sub3/page-chunks.json', pageChunks: '' },
      { name: 'com.application.demo.Card.srpk', jsonPath: 'Card/page-chunks.json', pageChunks: '' }
    ]
    const rpks = path.join(projectRoot, 'dist/com.application.demo.debug.1.0.0.rpks')
    const rpksBuf = fs.readFileSync(rpks)
    const rpksZipBuf = await JSZip.loadAsync(rpksBuf)
    for (let i = 0; i < pkgs.length; i++) {
      const pkg = pkgs[i]
      // 获取 rpks 里面的各个 rpk/srpk 的 buffer
      const pkgBuf = rpksZipBuf.file(pkg.name).async('nodebuffer')
      // 获取 rpk/srpk 里的各项文件 zip buffer
      const pkgZipBuf = await JSZip.loadAsync(pkgBuf)
      const pageChunks = pkgZipBuf.file(pkg.jsonPath)
      if (pkg.jsonPath === 'Card/page-chunks.json') {
        // 卡片端内无page-chunks.json
        expect(!pageChunks).toBeTruthy()
      } else {
        // 应用端内每个包都有page-chunks.json
        expect(!!pageChunks).toBeTruthy()
        let pageChunksJson = await pageChunks.async('string')
        // 仅记录key值即可
        pageChunksJson = JSON.stringify(JSON.parse(pageChunksJson), (key, val) => {
          return key ? '' : val
        })
        pkg.pageChunks = pageChunksJson.replace(/\\/gm, '/')
      }
    }
    expect(pkgs).toMatchSnapshot()
  }, 50000)

  afterAll(async () => {
    await del([projectRoot])
  }, 50000)
})
