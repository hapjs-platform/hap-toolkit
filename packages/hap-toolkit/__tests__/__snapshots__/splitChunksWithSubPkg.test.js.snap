// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`split chunks on a project split chunks with subpackages 1`] = `
Array [
  "app-chunks.json",
  "base/page-chunks.json",
  "Card/CardDemo1/index.css.json",
  "Card/CardDemo1/index.template.json",
  "Card/CardDemo2/index.css.json",
  "Card/CardDemo2/index.template.json",
  "manifest.json",
  "page-chunks.json",
  "Sub2/page-chunks.json",
  "Sub3/page-chunks.json",
]
`;

exports[`split chunks on a project split chunks with subpackages 2`] = `
Array [
  Object {
    "jsonPath": "app-chunks.json",
    "name": "com.application.demo.rpk",
    "pageChunks": "{\\"modules/modAll.js\\":\\"\\"}",
  },
  Object {
    "jsonPath": "page-chunks.json",
    "name": "com.application.demo.rpk",
    "pageChunks": "{\\"modules/modSubAll.js\\":\\"\\",\\"modules/modSub12.js\\":\\"\\",\\"modules/modSub2.js\\":\\"\\",\\"modules/modSub1.js\\":\\"\\"}",
  },
  Object {
    "jsonPath": "page-chunks.json",
    "name": "com.application.demo.base.srpk",
    "pageChunks": "{\\"modules/modSubAll.js\\":\\"\\",\\"modules/modSub12.js\\":\\"\\",\\"modules/modSub1.js\\":\\"\\"}",
  },
  Object {
    "jsonPath": "app-chunks.json",
    "name": "com.application.demo.base.srpk",
    "pageChunks": "{\\"modules/modAll.js\\":\\"\\"}",
  },
  Object {
    "jsonPath": "Sub2/page-chunks.json",
    "name": "com.application.demo.sub2.srpk",
    "pageChunks": "{\\"modules/modSubAll.js\\":\\"\\",\\"modules/modSub12.js\\":\\"\\",\\"modules/modSub2.js\\":\\"\\"}",
  },
  Object {
    "jsonPath": "Sub3/page-chunks.json",
    "name": "com.application.demo.sub3.srpk",
    "pageChunks": "{\\"modules/modSubAll.js\\":\\"\\",\\"modules/modSub12.js\\":\\"\\",\\"modules/modSub2.js\\":\\"\\",\\"modules/modSub1.js\\":\\"\\"}",
  },
  Object {
    "jsonPath": "Card/page-chunks.json",
    "name": "com.application.demo.Card.srpk",
    "pageChunks": "",
  },
]
`;
